### ISSUE TITLE
Brand new cluster out of space on first image pull

### ISSUE THREAD

**Issue by evanphx** [kubernetes/kubernetes#13174]
My first pod of a brand new cluster on AWS is dead in the water because of a disk space issue. I'm not immediately sure if this is docker or kubernetes fault, but I'm starting here.

The output from `kubectl describe pod catalog` is:

```
  Tue, 25 Aug 2015 19:48:54 -0700	Tue, 25 Aug 2015 19:48:54 -0700	1	{scheduler }						scheduled	Successfully assigned catalog to ip-172-20-0-200.us-west-2.compute.internal
  Tue, 25 Aug 2015 19:48:54 -0700	Tue, 25 Aug 2015 19:53:25 -0700	28	{kubelet ip-172-20-0-200.us-west-2.compute.internal}	failedMount	Unable to mount volumes for pod "catalog_default": mkdir /mnt/ephemeral/kubernetes/kubelet/pods/fda24eef-4b9c-11e5-a950-0239d486b295/volumes/kubernetes.io~secret/default-token-tru9t: no space left on device
  Tue, 25 Aug 2015 19:48:54 -0700	Tue, 25 Aug 2015 19:53:25 -0700	28	{kubelet ip-172-20-0-200.us-west-2.compute.internal}	failedSync	Error syncing pod, skipping: mkdir /mnt/ephemeral/kubernetes/kubelet/pods/fda24eef-4b9c-11e5-a950-0239d486b295/volumes/kubernetes.io~secret/default-token-tru9t: no space left on device
```

I ssh'd into the node and, sure enough, it's out of disk space. Here is the output from df:

```
ubuntu@ip-172-20-0-200:~$ df
Filesystem                           1K-blocks    Used Available Use% Mounted on
udev                                   1922840       0   1922840   0% /dev
tmpfs                                   385376   39420    345956  11% /run
/dev/xvda1                             8115168 1401996   6277896  19% /
tmpfs                                  1926864     560   1926304   1% /dev/shm
tmpfs                                     5120       0      5120   0% /run/lock
tmpfs                                  1926864       0   1926864   0% /sys/fs/cgroup
/dev/mapper/vg--ephemeral-docker       4038720 1410012   2403840  37% /mnt/ephemeral/docker
/dev/mapper/vg--ephemeral-kubernetes    194241  190145         0 100% /mnt/ephemeral/kubernetes
tmpfs                                   385376       0    385376   0% /run/user/1000
```

It appears that docker is using `aufs` as the driver (as I requested), so I'm lead to believe that it's kubernetes setting up these lvm volumes? And setting them up.. badly?

/cc @justinsb

**Comment by evanphx** [comment]
Welp, I know why that mount point is full:

```
root@ip-172-20-0-200:/mnt/ephemeral/kubernetes/kubelet/pods/d2c7debd-4b71-11e5-a950-0239d486b295/volumes/kubernetes.io~empty-dir# du -sh *
183M	influxdb-persistent-storage
```

I didn't disable the by default influxdb, etc monitoring and it appears that it has eaten 100% of the allocated space for empty-dir type volumes?

Which brings up a really good question: empty-dir volumes have an extremely low size limit?

/cc @pauldix

**Comment by justinsb** [comment]
Sorry.  This makes sense, and I actually just hit the same thing today.  (Which makes we wonder if this is somehow new behaviour, although I normally run btrfs and this is specific to not-btrfs, so it could just be a coincidence).

Here's the relevant code: https://github.com/kubernetes/kubernetes/blob/master/cluster/aws/templates/format-disks.sh#L129-L152

So in the aufs case we set up LVM to combine all your volumes, and allocate 95% to Docker and 5% to kubernetes storage.  Kubernetes is mostly logs and similar things, so 5% is plenty... Except that influxdb & elasticsearch use empty volumes, and these are mounted on the kubernetes area.

These therefore fill up rapidly.  I will ponder what should be done here.  We probably shouldn't be storing this data on ephemeral drives in the first place, but equally maybe we can just communicate a limit (even if it would effectively render influxdb & elasticseach useless).  Maybe we can somehow have a more dynamic allocation - although note that the docker area was itself 37% full, so it isn't the case that it is _ridiculously_ misallocated.

Suggestions very welcome!

**Comment by evanphx** [comment]
Seems like empty-dir volumes shouldn't be in the ephemeral at all. Especially since they can (and will) cause kubernetes to lose functionality over it's own ephemeral dir if there is a space issue. And I doubt you want to use quotas or something to limit the size of an empty-dir within the ephemeral mount.

If you want to limit the size of empty-dir's best to make a new volume (perhaps configured at node create time) and put them all in there. That means the pods compete for space in the new empty-dir allocation volume, but at least the users have control over it's size.

**Comment by evanphx** [comment]
@justinsb Is there anything I can do to work around this for now? Should I edit `format-disks.sh` and rebuild the cluster? Is there an easy way to control where empty-dir's go?

**Comment by evanphx** [comment]
Hm, seems like empty-dir's are perfect for ephemeral though.... :( Really, it's depends on usage. If a pod wants an empty-dir to store up to 1MB, it's perfect. But it's usage like influxdb's that break it entirely. Maybe empty-dir needs an option that specifies the max size and kubernetes can make a decision about where to put it based on that request?

**Comment by evanphx** [comment]
Looking through the `format-disks.sh` I see that the situation is avoided entirely if the instance has no ephemeral volumes, so I'm tearing down the cluster and rebuilding it with `m4`s instead, since they don't have ephemeral storage.

**Comment by evanphx** [comment]
Another data point. I rebuilt the cluster using node types that don't have ephemeral storage to avoid the above issue.

Well, that just delayed the issue. I had a pod not start properly and spew logs on a continual basis while I worked on fixing it. Those logs were in turn sent to logstash and elasticsearch, resulting in ES filling up /:

```
/dev/xvda1      7.8G  7.8G     0 100% /
```

With 2.9G from es:
```
root@ip-172-20-0-45:/var/lib/kubelet/pods/7ff5efe9-4bab-11e5-95ac-02359ee6514f/volumes/kubernetes.io~empty-dir# du -sh *
2.9G	es-persistent-storage
```

I'm going to consider the building logging facilities unstable at the moment because they seem to just cascade out of control and bring down functionality far too easily.

**Comment by justinsb** [comment]
This is the same issue as #13508 (commenting to link the two issues)

**Comment by romanek-adam** [comment]
I also encountered this issue. The questions is:

1. does this issue affect core k8s or end-user services or it is "just" logging?
2. depending on the answer to the question above... how serious is it? (for me it looks like a show-stopper for production use of k8s)
3. how is influxdb's persistent storage going to be managed in the long run?
4. is the data kept in it going to be erased periodically as to avoid filling up disk space? (just as regular logs are rotated and then deleted)
5. is there any workaround? (looks like using EC2 machines without ephemeral storage is not an option, according what's been reported by @evanphx)
5. what are the plans on solving this issue?

**Comment by justinsb** [comment]
We have a problem specific to AWS with AUFS, which is that I (erroneously) created a very small partition backing this, because I thought it was just k8s logs.

I fixed that in #13803 and just proposed a cherry-pick to 1.0 as #14272.

This will fix the fact that it runs out of disk almost immediately on AWS.

There are still some other issues, but I'd like to split them out into focused tickets:

I opened #14274 to track/discuss the fact that this data probably should be put onto a persistent volume anyway.  

I also opened #14276 to figure out how to make sure that we always have a sensible sized ephemeral storage on EBS (the backing for host volumes), no matter what instance type we choose.

I propose closing this issue, but mostly so we can focus on the particular tickets.  Feel free to discuss further here / reopen if you think I've missed an aspect.
